### PR TITLE
Add implicit pointer grabs in Wayland.

### PR DESCRIPTION
**Pull request by QBobWatson**
A Wayland client expects to receive pointer motion events from the time a button is pressed until it is released, even if the pointer leaves the client's surface while the button is held.  This is crucial for scrollbar functionality: a scrollbar should continue to drag even if the pointer leaves the window.  This does not currently work in qtile.  For instance, in Chrome, if you drag the scrollbar, move the mouse off the Chrome window, release the button, then move the pointer back onto the Chrome window, Chrome will resume scrolling without any button press.

In X11, this functionality is implemented on the client side using XGrabPointer or XIGrabDevice.  Wayland provides no such mechanism: this behavior must be implemented in the compositor.  This does not seem to be well documented.  However, if you look in gdk_wayland_device_grab in the gtk source, you'll see that grabbing the pointer in GDK is basically a no-op: it does some internal accounting and changes the cursor, then simply assumes the pointer is already grabbed.

Implicit grabs are implemented in Sway (see
sway/sway/input/seatop_{default,down}.c), although that's not what they're called.  In short, pressing a button on a client surface switches the cursor handling code into "down" mode, which short-cuts most of the WM pointer processing logic and sends events directly to the client, until the button is released.  (This is part of a more general mechanism in Sway: pointer events can be processed in several modes, e.g. for moving floating windows.)

This commit implements basic implicit grabs in qtile Wayland, following what Sway does (functionally, not structurally).  Pressing a button creates an ImplicitGrab object in the Core; releasing the button destroys it.  While the ImplicitGrab object is alive, all pointer events (button, {absolute_,}motion, axis) are sent directly to wlroots, bypassing qtile's event processing.  This seems to produce the desired UI behavior.

Implicit grabs should also be implemented for touch and tablet tool events, if and when those are added to qtile.

### CODE REVIEW

**Comment by QBobWatson** [comment]
> Thanks a lot for this and the detailed explanation, I will test and review this soon. So I guess this will fix #4096?

Yes, I believe so -- at least, now I can select text as described in the issue.

File: libqtile/backend/wayland/core.py [hunk 1]
```diff
@@ -587,8 +647,20 @@ def _on_cursor_button(self, _listener: Listener, event: pointer.PointerButtonEve
             handled = self._process_cursor_button(button, pressed)
 
         if not handled:
+            if self._pressed_button_count == 1 and found:
+                assert self._implicit_grab is None
```
Comment (m-col) [inline libqtile/backend/wayland/core.py:651]: Was this needed? Im surprised mypy wouldn't be able to infer this must be the case beyond the if statement on line 636

File: libqtile/backend/wayland/core.py [hunk 2]
```diff
@@ -108,6 +108,39 @@
     from libqtile.core.manager import Qtile
 
 
+class ImplicitGrab:
```
Comment (m-col) [inline libqtile/backend/wayland/core.py:111]: Might be worth using `HasListeners` just to follow the style of other classes that uses listeners?

File: libqtile/backend/wayland/core.py [hunk 3]
```diff
@@ -108,6 +108,39 @@
...
+        start_y: int,
+        start_sx: int,
+        start_sy: int,
+    ) -> None:
+        self.core = core
+        self.surface = surface
+        self.start_x = start_x
+        self.start_y = start_y
+        self.start_sx = start_sx
+        self.start_sy = start_sy
```
Comment (m-col) [inline libqtile/backend/wayland/core.py:133]: It looks like the two x values and two y values are only used together, so maybe we could collapse them. E.g.:
```suggestion
        self.start_dx = start_sx - start_x
        self.start_dy = start_sy - start_y
```
Then just use that one below in `_implicit_grab_motion`

File: libqtile/backend/wayland/core.py [hunk 4]
```diff
@@ -587,8 +647,20 @@ def _on_cursor_button(self, _listener: Listener, event: pointer.PointerButtonEve
...
 
         if not handled:
+            if self._pressed_button_count == 1 and found:
+                assert self._implicit_grab is None
+                win, surface, sx, sy = found
+                if surface:
+                    self._create_implicit_grab(event.time_msec, surface, sx, sy)
             self.seat.pointer_notify_button(event.time_msec, event.button, event.button_state)
 
+    def _implicit_grab_motion(self, time: int):
```
Comment (m-col) [inline libqtile/backend/wayland/core.py:657]: This may need a `-> None` return annotation

**Review by m-col — COMMENTED** [review]
This is really excellent work, thank you for this!

I've made a couple comments but overall this looks great, super clear and understandable.

File: libqtile/backend/wayland/core.py (hunk 1 above)
Comment (QBobWatson) [reply libqtile/backend/wayland/core.py:651]: Good point.  I originally had that line above line 636, but now it's automatic.

File: libqtile/backend/wayland/core.py (hunk 2 above)
Comment (QBobWatson) [reply libqtile/backend/wayland/core.py:111]: Sure, if you prefer.

File: libqtile/backend/wayland/core.py (hunk 3 above)
Comment (QBobWatson) [reply libqtile/backend/wayland/core.py:133]: Good point!  I was just copying the computation from Sway.

File: libqtile/backend/wayland/core.py (hunk 4 above)
Comment (QBobWatson) [reply libqtile/backend/wayland/core.py:657]: Done.  I'll push another commit shortly.

**Review by m-col — APPROVED** [review]
Works like a charm. Much appreciated :)

**Comment by ftpd** [comment]
Can please someone re-check that? It broke the ability to move tabs right/left in Firefox using the mouse, I'm also unable to move tickets on kanban board inside Jira ;-)
@m-col I'm tagging you directly as you were the one who merged this. Should I open a new issue?

**Pull request merged**